Add Pressure Gradient design document (#226)
E3SM-Project/Omega: components/omega/doc/design/PGrad.md

Reviewer: I don't think we are ready to use the Seamount test yet.  We can leave this but it would mean following up with that test after the pressure-gradient PR and other related changes go in.

Author: Let's leave it in and be sure to follow up on it.

(omega-design-pressure-grad)=
# Pressure Gradient

## 1 Overview
The pressure gradient will be responsible for computing the horizontal gradients of both the pressure and geopotential terms for the non-Boussinesq primative equations implemented in Omega.

> In the non-Boussinesq model, the vertical coordinate will be pressure as opposed to height.

Reviewer 2: Never mind. There is a link without parentheses. use [de Szoeke and Samelson 2002](https://journals.ametsoc.org/view/journals/phoc/32/7/1520-0485_2002_032_2194_tdbtba_2.0.co_2.xml) `https://journals.ametsoc.org/view/journals/phoc/32/7/1520-0485_2002_032_2194_tdbtba_2.0.co_2.xml`

In a pure pressure coordinate the pressure gradient term disappears (since the pressure does not vary along lines of constant pressure), just as how the geopotential term disappears in a pure z coordinate model.
However, similar to MPAS-Ocean's support for tilted height coordinates, Omega will allow for tilted pressure coordinates.

> This means that Omega will need to compute both the pressure and geopotential gradients.

Reviewer 3: I think in several places it says the pressure gradient will compute the geopotential gradient as well, but then I don't see that as an output in the Design section below... am I missing something?

Author: I had `Geopotential` as an input to the `computePressureGrad` method with the thought it would compute both gradients and add them to `Tend` (MPAS-equivalent of `pressure_and_zmid`). I can certainly break this up into separate functions if that would be preferable.

Author: I added a line in Section 4.2.4 making it more explicit that `computePressureGrad` accumulates the sum of the pressure and geopotential gradients in `Tend`

### 2.2 Requirement: Initial support for a simple centered pressure gradient

For initial global cases without ice shelf cavities, the pressure and geopotential gradients will be computed with a simple centered difference approximation. In later versions of Omega, one ore more  high-order pressure gradients will be implemented and will replace the centered approach in production runs.
However, the centered pressure gradient will remain an option for use in idealized testing.

### 2.3 Requirement: Flexibility to support a high-order pressure gradient
The centered pressure gradient will be insufficient for future versions of Omega that include ice shelf cavities and high resolution shelf breaks.
The pressure gradient framework should be flexible enough to support a high-order pressure gradient in the future.

### 2.4 Requirement: Flexibility to support tidal forcing and sea level change
In later versions of Omega, the pressure gradient will need to be able to include tidal forcing in the geopotential term.
These tidal forcings include both the tidal potential and the self attraction and loading terms.

> Additionally, other changes to the geoid

Author: Definitely a fragment. Thanks for catching this.

### 2.5 Requirement: Pressure gradient for barotropic mode

For split barotropic-baroclinic timestepping, the pressure gradient should provide the bottom pressure gradient tendency in the barotropic mode.

### Desired:

## 3 Algorithmic Formulation
The non-Boussinesq momentum equation is

> $$ \frac{D \mathbf{u}_h}{D t } + f\boldsymbol{k}\times \mathbf{u}_h + \left(v\nabla_A p + \nabla_A \phi \right) = \boldsymbol{\mathcal{F}}. $$

Author: Sorry about the rendering. I was having trouble with that too. It seemed to show up fine when I build the docs. I'll try an extra space to see if that helps.

> $$ \frac{D \mathbf{u}_h}{D t } + f\boldsymbol{k}\times \mathbf{u}_h + \left(v\nabla_A p + \nabla_A \phi \right) = \boldsymbol{\mathcal{F}}. $$

Author: And yes, it would be good to be consistent with the governing equation design doc.

> $$ \frac{D \mathbf{u}_h}{D t } + f\boldsymbol{k}\times \mathbf{u}_h + \left(v\nabla_A p + \nabla_A \phi \right) = \boldsymbol{\mathcal{F}}. $$

Reviewer 2: I think you can start with the governing equations DD [Final Layered Equations](https://portal.nersc.gov/project/e3sm/mpeterse/omega/design-doc-v1-eqns/_build/html/design/OmegaV1GoverningEqns.html#final-layered-equations) and match the notation there. For example, I used `r` rather than `A` for the layer-normal direction. It makes sense for individual design documents to review any derivations relevant to this design, but you could hide the other terms and just refer the reader to the governing equations DD. 

BTW you can link to another DD section in markdown with e.g. `[text](OmegaV0ShallowWater.md#thickness-and-tracer-terms)`

> $$ \frac{D \mathbf{u}_h}{D t } + f\boldsymbol{k}\times \mathbf{u}_h + \left(v\nabla_A p + \nabla_A \phi \right) = \boldsymbol{\mathcal{F}}. $$

Author: I removed the duplication of the governing equations and linked to the V1 design document.

where $\mathbf{u}_h$ is the horizontal velocity, $f$ is the Coriolis parameter, $v = \frac{1}{\rho}$ is the specific volume, $\rho = \rho(T,S,p)$ is the density, $p$ is the hydrostatic pressure, $\phi$ is the geopotential, and $\boldsymbol{\mathcal{F}}$ are the dissipative terms.
The operator $\nabla_A$ is the gradient along a constant surface, $A$, and the total derivative is

$$ \frac{D \mathbf{u}_h}{D t } = \left( \frac{\partial}{\partial t} \right)_A  + \mathbf{u}_h\cdot \nabla_A + \omega\frac{\partial}{\partial A}, $$

where $\omega$ is the cross coordinate flow.
In the layered non-Boussinesq equations, the prognostic variable is the pressure thickness $h_k$, so that the geometric thickness (in meters) is a diagnostic variable defined as:

> $$ \Delta z_k = v_k h_k. $$

Reviewer: I think this is probably my mistake. I think a factor of g is probably missing.

The pressure at vertical cell interfaces is the found by summing the pressure thicknesses:

> $$ p_{K+1/2} = p_{surf} + g\sum_{k=1}^K h_k. $$

Reviewer: And here I think I already mentioned that I think g should not be here.

Author: I believe it should if we define $$h = \rho \Delta z$$.

> The geopotential at vertical cell integrates is found by summing the pressure thicknesses multiplied by the specific volume:

Reviewer 4: ```suggestion
The geopotential at vertical cell interfaces is found by summing the pressure thicknesses multiplied by the specific volume:
```

Author: thanks for catching this!

> $$ \phi_{K+1/2} = g\left(z_b + \sum_{k=K}^{N} v_k h_k\right), $$

Reviewer 2: I think a lower case `k` and primed summation index `k'` is more standard. Upper case `K` is often used for a maximum index. But more generally, the important diagnostic variable is mid-level pressure, so I think that should be written here. See my pressure equation in the [Discrete Equation section](https://portal.nersc.gov/project/e3sm/mpeterse/omega/design-doc-v1-eqns/_build/html/design/OmegaV1GoverningEqns.html#discrete-equations).

where $z_b$ is the (positive-up) bottom depth.

The discrete gradient operator at an edge is:

$$ \nabla {(\cdot)} = \frac{1}{d_e} \sum_{i\in CE(e)} -n_{e,i} (\cdot)_i $$

where $d_e$ is the distance between cell centers, $CE(e)$ are the cells on edge $e$, and $n_{e,i}$ is the sign of the edge normal with respect to cell $i$.
Therefore the centered pressure gradient will be calculated as:

$$ T^p_k = \frac{1}{d_e} \left( \widehat{v}_{k,e} \sum_{i \in CE(e)} n_{e,i} \overline{p}_{k,i} + \sum_{i\in CE(e)} n_{e,i} \overline{\phi}_{k,i}\right), $$

$$ = \frac{1}{d_e} \left(  \sum_{i \in CE(e)} n_{e,i} \left( \widehat{v}_{k,e}\overline{p}_{k,i} + \overline{\phi}_{k,i} \right) \right), $$

with the vertical averaging operator defined as:

> $$ \overline{(\cdot)} = \frac{1}{2} \left((\cdot)_{k+1/2} + (\cdot)_{k-1/2} \right)$$

Reviewer 2: Let's use mid-depth as the "native" location, so no bars or subscripts. Vertical interfaces can use $k+1/2$ or $top$ and $bot$. See notation in the [Discrete Equations section](https://portal.nersc.gov/project/e3sm/mpeterse/omega/design-doc-v1-eqns/_build/html/design/OmegaV1GoverningEqns.html#discrete-equations).

and the horizontal averaging operator:

> $$\widehat{(\cdot)} = \frac{1}{2} \left( (\cdot)_{i=1} + (\cdot)_{i=2}\right)$$

Reviewer 2: For uniformity, see notation in [Discrete Equations](https://portal.nersc.gov/project/e3sm/mpeterse/omega/design-doc-v1-eqns/_build/html/design/OmegaV1GoverningEqns.html#discrete-equations). Each variable has a native location (cell, edge, vertex) denoted by subscript (i,e,v). I used square brackets with a subscript to denote a horizontal average, like this:
![image](https://github.com/user-attachments/assets/dddfbba1-97bd-4bc8-bd7a-82b6d776dae8)
These are defined with that notation in the [V0 doc](https://docs.e3sm.org/Omega/omega/develop/design/OmegaV0ShallowWater.html#cell-to-edge-interpolation).

Reviewer 2: Also, the $i$ in your equation are not specified. The better notation is 
![image](https://github.com/user-attachments/assets/c8aa59e2-2bee-4f11-8b46-42f5ff74ca5a)

## 4 Design

### 4.1 Data types and parameters

The `PressureGradient` class will be used to perform the horizontal gradients of the pressure and geopotential
```c++
class PressureGrad{
    public:
    private:
        PressureGradCentered CenteredPGrad;

>         PressureGradHighOrder HighOrderPGrad;

Reviewer 4: since not being implemented now do we want to make a note it is something to come later somewhere?  May be more appropriate for one of the guides.

Author: sure, I can make a comment here and draw attention to that in the guides.

        PressureGradType PressureGradChoice;
        I4 NVertLevels;
        I4 NChuncks;
        Array2DI4 CellsOnEdge;
        Array1DReal DvEdge;
        Array1DReal EdgeSignOnCell;
}
```
The user will select a pressure gradient option at runtime in the input yaml file under the pressure gradient section:
```yaml
    PressureGrad:
       PressureGradType: 'centered'
```
An `enum class` will be used to specify options for the pressure gradient used for an Omega simulation:
```c++
enum class PressureGradType{
   Centered,
   HighOrder
}
```
The functions to compute the centered and high order pressure gradient terms will be implemented as functors and the pressure gradient class will have private instances of these classes.
### 4.2 Methods

```c++
class PressureGrad{
    public:
        static PressureGrad *create();
        void computePressureGrad();

>         void computePressureGradBtr();

Reviewer 5: The Btr function is not described anywhere?

Author: Thanks, I can take care of that.

    private:

}
```
#### 4.2.1 Creation

The constructor will be responsible for storing any static mesh information as private variables and handling the selection of the user-specified pressure gradient option.
```c++
PressureGrad::PressureGrad(const HorzMesh *Mesh, int NVertLevels, Config *Options);
```

The create method will take the same arguments as the constructor plus a name.
It calls the constructor to create a new pressure gradient instance, and put it in the static map of all pressure gradients.
It will return a pointer to the newly created object.
```c++
PressureGrad *PressureGrad::create(const std::string &Name, const HorzMesh *Mesh, int NVertLevels, Config *Options);
```

#### 4.2.2 Initialization

The init method will create the default pressure gradient and return an error code:
```c++
int PressureGrad::init();
```

#### 4.2.3 Retrieval

There will be methods for getting the default and non-default pressure gradient instances:
```c++
PressureGrad *PressureGrad::getDefault();

> PressureGrad *PressureGrad::get(const std::string &Name);

Reviewer 5: What is the use case for multiple gradient instances?

Author: My thought was that we could use a separate instance for the baroclinic and barotropic pressure gradients. But perhaps this doesn't make much sense to do that.

Author: Also, I've just been using this as a template since it was established early on for a lot of the infrastructure classes (when it made sense to support multiple instances). Now that may not be needed for Omega-1, is there a convention we want to use to do the retrieval for a single instance between the `init` and `run` routines? This is something we were also wondering for the EOS (@alicebarthel)

Reviewer 5: It's just like the multiple instance - you just keep the managed (unique) pointer to the single instance rather than saving a map of multiple instances (see for example the single calendar in TimeMgr). And it's fine to have a single instance of the baroclinic and a single instance of barotropic with a get for each.  But a retrieval by name involves a string comparison and search through the map so adds some unnecessary overhead if we don't need it. It's one of the things that slowed down MPAS.

(omega-design-pressure-grad)=
# Pressure Gradient

## 1 Overview
The pressure gradient will be responsible for computing the horizontal gradients of both the pressure and geopotential terms for the non-Boussinesq primitive equations implemented in Omega.
In the non-Boussinesq model, the conserved quantity is mass rather than volume.

> In Omega the prognostic variable $\tilde{h}$ is a pseudo thickness, rather than geometric thickness in m as in a Boussinesq model.

Reviewer 6: But isn't the pseudo thickness also in m (i.e. units of length)? If so, it's confusing to mention "in m" for one but not the other.

>  Some non-Boussinesq models are written in pressure coordinates (e.g. [de Szoeke and Samelson 2002](https://journals.ametsoc.org/view/journals/phoc/32/7/1520-0485_2002_032_2194_tdbtba_2.0.co_2.xml).

Reviewer: Missing a second close parenthesis:
```suggestion
 Some non-Boussinesq models are written in pressure coordinates (e.g. [de Szoeke and Samelson 2002](https://journals.ametsoc.org/view/journals/phoc/32/7/1520-0485_2002_032_2194_tdbtba_2.0.co_2.xml)).
```

Author: thanks!

>  However, Omega is written in general vertical coordinates and can reference either pressure $p$ or distance $z$ in the vertical.

Reviewer: ```suggestion
 However, Omega is written in general vertical coordinates and can reference either pressure $p$ or height $z$ in the vertical.
```

In a pure pressure coordinate the pressure gradient term disappears (since the pressure does not vary along lines of constant pressure), just as how the geopotential term disappears in a pure z coordinate model.
However, similar to MPAS-Ocean's support for tilted height coordinates, Omega will allow for tilted pressure coordinates.
This means that Omega will need to compute both the pressure and geopotential gradients.

## 2 Requirements

### 2.1 Requirement: Support for tilted pressure coordinates for the non-Boussinesq primitive equations

> The pressure gradient will compute the horizontal gradients of both the pressure and geopotential to support tilted pressure coordinates in the non-Boussinesq model.

Reviewer: ```suggestion
The pressure gradient will compute the horizontal (along-layer) gradients of both the pressure and geopotential to support tilted pressure coordinates in the non-Boussinesq model.
```

> This will allow for the use of a $p^\star$ coordinate, which functions similarly to the $z^\star$ in the Boussinesq MPAS-Ocean model.

Reviewer 4: will this be updated to the \tilde{z} coordinate instead of $p^\star$?  I guess psuedo height could be thought of as $p^\star$ like so maybe no need to change anything

Reviewer: $\tilde{z}*$? 😝

Author: I was hoping we could continue calling it p* since \tilde{z}^* seems confusing.

### 2.2 Requirement: Initial support for a simple centered pressure gradient

For initial global cases without ice shelf cavities, the pressure and geopotential gradients will be computed with a simple centered difference approximation. In later versions of Omega, one or more  high-order pressure gradients will be implemented and will replace the centered approach in production runs.
However, the centered pressure gradient will remain an option for use in idealized testing.

### 2.3 Requirement: Flexibility to support a high-order pressure gradient
The centered pressure gradient will be insufficient for future versions of Omega that include ice shelf cavities and high resolution shelf breaks.
The pressure gradient framework should be flexible enough to support a high-order pressure gradient in the future.

> The high order pressure gradient will be similar to [Adcroft et al. 2008](https://doi.org/10.1016/j.ocemod.2008.02.001).

Reviewer: Let's hedge on this for now.
```suggestion
The high order pressure gradient may be similar to [Adcroft et al. 2008](https://doi.org/10.1016/j.ocemod.2008.02.001).
```

### 2.4 Requirement: Flexibility to support tidal forcing and sea level change
In later versions of Omega, the pressure gradient will need to be able to include tidal forcing in the geopotential term.
These tidal forcings include both the tidal potential and the self attraction and loading terms.
Additionally, other long-term changes to the geoid can be included in the geopotential.

### 2.5 Disired: Pressure gradient for barotropic mode

For split barotropic-baroclinic timestepping, the pressure gradient should provide the bottom pressure gradient tendency in the barotropic mode.
This will be added in a future version when split time stepping is implemented.

> 
> ### Desired:

Reviewer: ```suggestion
```

## 3 Algorithmic Formulation
### 3.1 Centered Pressure Gradient
In the layered non-Boussinesq [momentum equation](OmegaV1GoverningEqns.md#discrete-momentum) solved in Omega, the pressure gradient tendency term for edge $e$ and level $k$, $T^p_{e,k}$, includes the gradient of the pressure and the gradient of the geopotential,

$$
T^p_{e,k} = -\left[ \alpha_{i,k} \right]_e \nabla p_{i,k} - \nabla \Phi_{i,k},
$$

where the second term is necessary to account for tilted layers that occur when using a general vertical coordinate.

> In this equation, $\alpha_{i,k}$ is the specific volume for cell $i$ at the mid-point of level $k$, $p_{i,k}$ is the pressure, and $\Phi_{i,k}$ is the geopotential.

Reviewer 6: Can we specify that p and \phi are all mid-layer values?Just to be specific.

The discrete gradient operator at an edge is:

$$
 \nabla {(\cdot)} = \frac{1}{d_e} \sum_{i\in CE(e)} -n_{e,i} (\cdot)_i
$$

where $d_e$ is the distance between cell centers, $CE(e)$ are the cells on edge $e$, and $n_{e,i}$ is the sign of the edge normal with respect to cell $i$.

> The horizontal averaging operator is:

Reviewer 6: ```suggestion
The (cell-to-edge) horizontal averaging operator is:
```

$$
 [\cdot]_e = \frac{1}{2}\sum_{i\in CE(e)} (\cdot)_i
$$

Therefore, the centered pressure gradient will be calculated as:

$$
 T^p_{e,k} = \frac{1}{d_e} \left( [\alpha_{i,k}]_e \sum_{i \in CE(e)} n_{e,i} p_{k,i} + \sum_{i\in CE(e)} n_{e,i} \Phi_{k,i}\right),
$$

$$
 = \frac{1}{d_e} \left(  \sum_{i \in CE(e)} n_{e,i} \left( [\alpha_{i,k}]_ep_{k,i} + \Phi_{k,i} \right) \right),
$$

### 3.2 Barotropic Pressure Gradient

When split baroclinic-barotropic time stepping is implemented in the future, the barotropic pressure gradient will be calculated by the pressure gradient class.
The barotropic pressure gradient is found by depth integrating the pressure gradient.
The pressure is

$$
p(z) = p_b - g \int^z_{-h} \rho dz^\prime,

> $$

Reviewer: Why is $z$ and not $\tilde{z}$ the coordinate of integration in this whole section?  And is it worth retaining this section or should we address it when we address the barotropic equation as a whole?

Author: I've decided to leave the barotropic pressure gradient to be discussed in the split timestepping design doc.

where $p_b$ is the bottom pressure.
The bottom pressure is the sum of the atmospheric surface pressure, $p_s$, and the pressure contribution of the water column:

$$
p(z) &=  p_s + g\int_{-h}^\eta \rho dz - g \int^z_{-h} \rho dz^\prime, \\
     &=  p_s + g\rho_0\widetilde{H} - g \int^z_{-h} \rho dz^\prime,
$$

where the total water column pseudo height is expressed by

$$
\widetilde{H} = \int_{-h}^\eta \frac{\rho}{\rho_0} dz.
$$

$\widetilde{H}$ is the prognositc variable in the barotropic continuity equation.

> The vertical integral of the pressure gradient is

Reviewer: This is the vertical average, not the vertical integral, right?
```suggestion
The vertical average of the pressure gradient is
```

$$
\frac{1}{\rho_0\widetilde{H}}\int^\eta_{-h} \nabla p dz &= \frac{1}{\rho_0\widetilde{H}}\int^\eta_{-h} \nabla \left( p_s + g\rho_0 \widetilde{H} - g \int^z_{-h} \rho dz^\prime \right) dz, \\
                           &= \frac{H}{\rho_0\widetilde{H}}\nabla p_s + \frac{gH}{\widetilde{H}}\nabla \widetilde{H} - \frac{g}{\rho_0\widetilde{H}} \int_{-h}^\eta \left( \nabla \int_{-h}^z \rho dz^\prime\right) dz,
$$

where the height of the water column is represented by $H$.
The $1/\rho_0\widetilde{H}$ factor comes vertically integrating the material derivative and expressing the resulting barotropic momentum equation in non-conservative form.

> Therefore, the barotorpic pressure gradient term is discretized as:

Reviewer 6: ```suggestion
Therefore, the barotropic pressure gradient term is discretized as:
```

$$

> \overline{T}_e^p = g\left[ \frac{H_i}{\widetilde{H}_i} \right]_e\sum_{i \in CE(e)} n_{e,i}\widetilde{H}_e

Reviewer: How will we know $H_i$ in this equation?  This seems pretty unknowable in the barotropic equation to me.

Author: I've decided to leave the barotropic pressure gradient to be discussed in the split timestepping design doc.

$$

## 4 Design

### 4.1 Data types and parameters

The `PressureGradient` class will be used to perform the horizontal gradients of the pressure and geopotential
```c++
class PressureGrad{
    public:
    private:
        std::unique_ptr<PressureGrad> OmegaPressureGrad;
        PressureGradCentered CenteredPGrad;

>         PressureGradHighOrder HighOrderPGrad; // To be implemented later

Reviewer: I'm not confident that we will have one and only one higher-order pressure gradient.  We have several in MPAS-Ocean.  Assuming that this is just a placeholder for one or more future classes, can we state that explicitly in the text?

Author: Yes, I'll mention that.

```c++
void PressureGrad::computePressureGrad(const Array2DReal &Tend,
                                       const Array2DReal &Pressure,
                                       const Array2DReal &Geopotential,
                                       const Array2DReal &SpecVol) {
OMEGA_SCOPE(LocCenteredPGrad, CenteredPGrad)
OMEGA_SCOPE(LocHighOrderPGrad, HighOrderPGrad)
    if (PressureGradChoice == PressureGradType::Centered){
       parallelFor("pgrad-centered", {NCellsAll, NChunks},
       KOKKOS_LAMBDA(int ICell, int KChunk) {
          LocCenteredPGrad(Tend, Pressure, Geopotential, SpecVol);
       });
    }
    else if (PressureGradChoice == PressureGradType::HighOrder){
       parallelFor("pgrad-highorder", {NCellsAll, NChunks},
       KOKKOS_LAMBDA(int ICell, int KChunk) {
          LocHighOrderPGrad(Tend, Pressure, Geopotential, SpecVol);
        });
    }
```

#### 4.2.5 Destruction and removal

No operations are needed in the destructor.
The clear method will remove the pressure gradient instance, calling the destructor in the process.
```c++
void PressureGrad::clear();
```

## Verification and Testing

### Test: Spatial convergence to exact solution

> For a given analytical $v$,  $h$, and $b$, the spatial convergence of the pressure gradient can be assessed by computing errors on progressively finer meshes.

Reviewer: What is $b$ in this case?

Reviewer: I think maybe this should be fleshed out a bit more.  What analytic test would we use?

In a pure pressure coordinate the pressure gradient term disappears (since the pressure does not vary along lines of constant pressure), just as how the geopotential term disappears in a pure z coordinate model.
However, similar to MPAS-Ocean's support for tilted height coordinates, Omega will allow for tilted pressure coordinates.
This means that Omega will need to compute both the pressure and geopotential gradients.

## 2 Requirements

### 2.1 Requirement: Support for tilted pressure coordinates for the non-Boussinesq primitive equations

The pressure gradient will compute the horizontal gradients of both the pressure and geopotential to support tilted pressure coordinates in the non-Boussinesq model.
This will allow for the use of a $p^\star$ coordinate, which functions similarly to the $z^\star$ in the Boussinesq MPAS-Ocean model.
Note that we use the name "$p^\star$" to refer to the vertical coordinate, in Omega it will be expressed in terms of the pseudo-height, $\tilde{z}$, as opposed to pressure directly.

### 2.2 Requirement: Initial support for a simple centered pressure gradient

> For initial global cases without ice shelf cavities, the pressure and geopotential gradients will be computed with a simple centered difference approximation. In later versions of Omega, one or more  high-order pressure gradients will be implemented and will replace the centered approach in production runs.

Reviewer 4: one or more seems different from the next requirement where Adcroft 2008 is mentioned.  Is the plan to leave open for multiple options?

Reviewer 4: the hedge Xylar proposed below seems like a good choice will->may and resolves the inconsistency

In later versions of Omega, the pressure gradient will need to be able to include tidal forcing in the geopotential term.
These tidal forcings include both the tidal potential and the self attraction and loading terms.
Additionally, other long-term changes to the geoid can be included in the geopotential.
This requirement is satisfied by tidal forcing terms being included in the geopotential calculation in the `VertCoord` class.

### 2.5 Desired: Pressure gradient for barotropic mode

For split barotropic-baroclinic timestepping, the pressure gradient should provide the bottom pressure gradient tendency in the barotropic mode.
The details of the barotropic pressure gradient will be added in a future design document for split time stepping.

## 3 Algorithmic Formulation
### 3.1 Centered Pressure Gradient
In the layered non-Boussinesq  {ref}`momentum equation <omega-v1-momentum-eq>` solved in Omega, the pressure gradient tendency term for edge $e$ and layer $k$, $T^p_{e,k}$, includes the gradient of the geopotential, the gradient of a term involving pressure, and two terms evaluated at the cell interface:

$$

> T^p_{e,k} = - \left(\nabla \Phi \right)_{e,k} - \frac{1}{\left[\tilde{h}_k\right]_e} \nabla \left( \tilde{h}_k \alpha_k p_k \right) + \frac{1}{\left[\tilde{h}_k\right]_e} \left\{ \left[ \alpha p \nabla \tilde{z}\right]_{e,k}^\text{top} -  \left[ \alpha p \nabla \tilde{z}\right]_{e,k+1}^\text{top} \right\}.

Reviewer 2: ```suggestion
T^p_{e,k} = - \left(\nabla \Phi \right)_{e,k} - \frac{1}{\left[\tilde{h}_k\right]_e} \nabla \left( \tilde{h}_k \alpha_k p_k \right) - \frac{1}{\left[\tilde{h}_k\right]_e} \left\{ \left[ \alpha p \nabla \tilde{z}\right]_{e,k}^\text{top} -  \left[ \alpha p \nabla \tilde{z}\right]_{e,k+1}^\text{top} \right\}.
```
Compare to last term in [V1 DD eqn 49](https://portal.nersc.gov/project/e3sm/mpeterse/omega/pgrad-design-251009/design/OmegaV1GoverningEqns.html#equation-layer-momentum-final-simple), which is negative. 
<img width="394" height="71" alt="image" src="https://github.com/user-attachments/assets/2a099b37-c4f7-4983-abe1-46c8e5af0b64" />
It looks like this is the only error, as the signs in the discrete form below are correct.

Author: Thanks for catching that @mark-petersen.

$$

> The geopotential and interface terms are necessary to account for tilted layers that occur when using a general vertical coordinate, where the gradient operator is taken along layers.

Reviewer: I'm find to leave this but just wanted to note that, for layers at constant pseudo-height (rather than geometric height), the geoponential gradient would still be non-zero as will the second term (unless density is constant along layers).

> In this equation, $\alpha_{i,k}$ specific volume, $p_{i,k}$ is the pressure, and $\Phi_{i,k}$ is the geopotential.

Reviewer: ```suggestion
In this equation, $\alpha_{i,k}$ is the specific volume, $p_{i,k}$ is the pressure, and $\Phi_{i,k}$ is the geopotential.
```

> These three quantities are evaluated at the mid-point of layer $k$ of cell $i$ in the first two terms and at the cell interfaces in the third term along with the interface psudo-height, \tilde{z}.

Reviewer: ```suggestion
These three quantities are evaluated at the mid-point of layer $k$ of cell $i$ in the first two terms and at the cell interfaces in the third term along with the interface psudo-height, $\tilde{z}$.
```

The discrete gradient operator at an edge is:

$$
 \nabla {(\cdot)} = \frac{1}{d_e} \sum_{i\in CE(e)} -n_{e,i} (\cdot)_i
$$

where $d_e$ is the distance between cell centers, $CE(e)$ are the cells on edge $e$, and $n_{e,i}$ is the sign of the edge normal with respect to cell $i$.
The (cell-to-edge) horizontal averaging operator is:

$$
 [\cdot]_e = \frac{1}{2}\sum_{i\in CE(e)} (\cdot)_i
$$

Therefore, the centered pressure gradient will be calculated as:

$$
 T^p_{e,k} = \frac{1}{d_e}\sum_{i\in CE(e)} n_{e,i} \Phi_{i,k} + \frac{2}{d_e\sum_{i \in CE(e)}\tilde{h}_{i,k}}\left(\sum_{i \in CE(e)} n_{e,i} \tilde{h}_{i,k}\alpha_{i,k}p_{i,k}\right. \\ \left. - \frac{1}{2} \sum_{i\in CE(e)} \alpha_{i,k-1/2} p_{i,k-1/2}\sum_{i\in CE(e)} n_{e,i}\tilde{z}_{i,k-1/2} \right.\\ \left.+ \frac{1}{2} \sum_{i\in CE(e)} \alpha_{i,k+1/2} p_{i,k+1/2}\sum_{i\in CE(e)} n_{e,i}\tilde{z}_{i,k+1/2}\right),
$$

> where $k-1/2$ and $k+1/2$ refer to the top and bottom of layer $k$, respectively.

Reviewer: It isn't clear in the design how $\alpha_{i,k+1/2}$ and $p_{i,k+1/2}$ will be calculated.

Reviewer: My feeling would be that $\alpha_{i,k+1/2} = 1/2 (\alpha_{i,k} + \alpha_{i,k+1})$.  But I would think $p_{i,k+1/2}$ would be known, since it is directly proportional to the pseudo-height that is known at both interfaces and layer midpoints.

Reviewer: As with the other products, we should discuss whether the right thing to do is compute the product at cell centers and layer midpoints, and then average to edges and interfaces.  In which case, we would first compute the product $\alpha_{i,k} p_{i,k}$ and then do the 4-point average to get it to the edge and layer interface.

Author: I'll update this as per our discussion today. Thanks!

& - \int_A \rho_0 \left[ \left< \alpha \right> \left<p \nabla \tilde{z}_k^{\text{top}} \right> + \left<\alpha^\prime \left(p \nabla \tilde{z}_k^{\text{top}}\right)^\prime\right> \right]_{\tilde{z} = \tilde{z}_k^{\text{top}}} \, dA \\
& + \int_A \rho_0 \left[ \left< \alpha \right> \left<p \nabla \tilde{z}_k^{\text{bot}} \right> + \left<\alpha^\prime \left(p \nabla \tilde{z}_k^{\text{bot}}\right)^\prime\right> \right]_{\tilde{z} = \tilde{z}_k^{\text{bot}}} \, dA.
$$
To obtain the expression that will be used, we neglect the turblent correlations and drop the Reynold's average notation for single variables:
$$
T^p &= - \int_A \int_{\tilde{z}_k^{\text{bot}}}^{\tilde{z}_k^{\text{top}}} \rho_0 \, \left( \nabla \Phi \right) \, d\tilde{z} \, dA \\
& - \int_{\partial A} \left( \int_{\tilde{z}_k^{\text{bot}}}^{\tilde{z}_k^{\text{top}}} \rho_0 \left(\alpha p \right) \, d\tilde{z} \right) dl \\
& - \int_A \rho_0 \left[ \alpha \left<p \nabla \tilde{z}_k^{\text{top}} \right> \right]_{\tilde{z} = \tilde{z}_k^{\text{top}}} \, dA \\
& + \int_A \rho_0 \left[ \alpha \left<p \nabla \tilde{z}_k^{\text{bot}} \right> \right]_{\tilde{z} = \tilde{z}_k^{\text{bot}}} \, dA.
$$

> These volume and area integrals will be computed using quadrature to account for the variablity of $\alpha$ with the recontructed values of temperature, salinity, and pressure atthe quadrature points.

Reviewer: ```suggestion
These volume and area integrals will be computed using quadrature to account for the variability of $\alpha$ with the reconstructed values of temperature, salinity, and pressure at the quadrature points.
```

             parallelForInner(Team, NChunks, [=](const int KChunk) {
                LocCenteredPGrad(Tend, IEdge, KChunk, Pressure, Geopotential, SpecVol);
             });
       });
    }
    else if (PressureGradChoice == PressureGradType::HighOrder){
       parallelForOuter(
          {NEdgesAll}, KOKKOS_LAMBDA(int IEdge, const TeamMember &Team) {
             const int NChunks = computeNChunks(MinLyrEdgeBottom, MaxLyrEdgeTop, IEdge);
             parallelForInner(Team, NChunks, [=](const int KChunk) {
                LocHighOrderPGrad(Tend, IEdge, KChunk, Pressure, Geopotential, SpecVol);
             });
       });
    }
```

#### 4.2.5 Destruction and removal

No operations are needed in the destructor.
The clear method will remove the pressure gradient instance, calling the destructor in the process.
```c++
void PressureGrad::clear();
```

## Verification and Testing

### Test: Spatial convergence to exact solution
For given analytical functions of $\alpha$, $h$, and $z$, the spatial convergence of the pressure gradient can be assessed by computing errors on progressively finer meshes.

### Test: Baroclinic gyre

> The baroclinic gyre test case will test the pressure gradient term in the full non-Boussinesq equations.

Reviewer: I don't think we'll have the baroclinic gyre test in time for our testing needs.

Let's link to the seamount and the overflow test cases here (actually liking to the Polaris docs).

And we'll add my 2-column test as soon as I have a draft PR.

Author: @xylar, I switched the baroclinic gyre out in favor of the overflow and added Polaris urls for the seamount and overflow cases.

This requirement is satisfied by tidal forcing terms being included in the geopotential calculation in the `VertCoord` class.

### 2.5 Desired: Pressure gradient for barotropic mode

For split barotropic-baroclinic timestepping, the pressure gradient should provide the bottom pressure gradient tendency in the barotropic mode.
The details of the barotropic pressure gradient will be added in a future design document for split time stepping.

## 3 Algorithmic Formulation
In the layered non-Boussinesq equation solved in Omega, the pressure and geopotential gradient terms are taken along a generalized coordinate $r$:
$$
\alpha \nabla_z p + \nabla_z \Phi = \alpha\left( \nabla_r p - \frac{\partial r}{\partial z}\frac{\partial p}{\partial r}\nabla_r z\right) + \nabla_r \Phi - \frac{\partial r}{\partial z}\frac{\partial \Phi}{\partial r} \nabla_r z.
$$
Using the hydrostatic relation
$$
\frac{\partial p}{\partial z} &= - \rho g, \\
\frac{\partial p}{\partial r}\frac{\partial r}{\partial z} &= - \rho g, \\
\frac{\partial r}{\partial z} &= -\rho g \frac{\partial r}{\partial p}.
$$
to subtitute for $\partial r/\partial z$, and substituting $\Phi = gz + \phi_{TP} + \phi_{SAL}$ yields:
$$

> \alpha \nabla_z p &= \alpha\left( \nabla_r p + \rho g \nabla_r z\right) + \nabla_r \Phi - \frac{\partial r}{\partial z}\frac{\partial gz}{\partial r} \nabla_r z, \\

Reviewer 7: Found a typo while reviewing #353 
```suggestion
\alpha \nabla_z p + \nabla_z \Phi &= \alpha\left( \nabla_r p + \rho g \nabla_r z\right) + \nabla_r \Phi - \frac{\partial r}{\partial z}\frac{\partial gz}{\partial r} \nabla_r z, \\
```
